Client/WebFilters/CosmeticFiltersResourceDownloader.swift

```diff
@@ -0,0 +1,271 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Shared
+import BraveShared
+
+private let log = Logger.browserLogger
+
+private struct CosmeticFilterNetworkResource {
+    let resource: CachedNetworkResource
+    let fileType: FileType
+    let type: CosmeticFiltersResourceDownloader.CosmeticFilterType
+}
+
+class CosmeticFiltersResourceDownloader {
+    static let shared = CosmeticFiltersResourceDownloader()
+    
+    private let networkManager: NetworkManager
+    
+    static let folderName = "cmf-data"
+    private let servicesKeyName = "SERVICES_KEY"
+    private let servicesKeyHeaderValue = "BraveServiceKey"
+    private var engine = AdblockRustEngine()
+    
+    static let endpoint = "https://raw.githubusercontent.com/iccub/brave-ios/development"
+    
+    private init(networkManager: NetworkManager = NetworkManager()) {
+        self.networkManager = networkManager
+    }
+    
+    /// Initialized with year 1970 to force adblock fetch at first launch.
+    private(set) var lastFetchDate = Date(timeIntervalSince1970: 0)
+    
+    func startLoading() {
+        let now = Date()
+        let fetchInterval = AppConstants.buildChannel.isPublic ? 6.hours : 10.minutes
+        
+        if now.timeIntervalSince(lastFetchDate) >= fetchInterval {
+            lastFetchDate = now
+            
+            // MUST re-create the engine otherwise we get insane load times when calling:
+            // `engine_add_resources`
+            // This is because `engine_add_resources` will ADD resources, and not delete old ones
+            // Thus we get a huge amount of memory usage and slow down.
+            engine = AdblockRustEngine()
+            
+            loadDownloadedFiles()
+            downloadCosmeticSamples()
+            downloadResourceSamples()
```

Comment:
Forgive me if I'm misunderstanding the control flow here, but I think the engine should not be recreated until all the new resources have been successfully downloaded. Otherwise users with slow or inconsistent network connections might start getting occasional page loads with unblocked content.

In theory a slow network connection should also prevent page resources from loading, but there's potential for network race conditions, and the cosmetic filters and scriptlets will no longer be applied to cached pages in the meantime.

Reply:
Fixed. I download stuff and add it only to a new engine, then replace the old one with the new one once completed.